xray/segment.go

```diff
@@ -164,15 +158,44 @@ func (seg *Segment) RemoveSubsegment(remove *Segment) bool {
 	return false
 }
 
+func (seg *Segment) handleContextDone() {
+	seg.Lock()
+	defer seg.Unlock()
+
+	seg.ContextDone = true
+	if !seg.InProgress && !seg.Emitted {
+		seg.flush()
+	}
+}
+
+func (seg *Segment) flush() {
```

Comment:
hmm, i feel some FUD about assuming that the `seg.Lock()` is held by all callers of this function. this might just be my inexperience.

Reply:
understandable, but the fix here was to trust callers to lock properly. golang doesn't seem to have the concept of reentrant locks (which is what we need here).